### PR TITLE
DM-32842: `open` and miscelleneous minor improvements

### PR DESCRIPTION
**Pull request by TallJimbo**
## Checklist

- [ ] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: tests/test_uri.py [hunk 1]
```diff
@@ -483,7 +566,6 @@ def testTemporary(self):
             self.assertFalse(tmp.exists(), f"uri: {tmp}")
             tmp.write(b"abcd")
             self.assertTrue(tmp.exists(), f"uri: {tmp}")
-            self.assertTrue(tmp.isTemporary)
```
Comment (timj) [inline tests/test_uri.py:486]: Why was this deleted?

**Review by timj — APPROVED** [review]
Looks great. Thanks for doing this.

It does remind me that we need to integrate the ButlerURI patching added in lsst/pipe_base#163.

File: python/lsst/resources/_resourcePath.py [hunk 2]
```diff
@@ -24,9 +26,9 @@
 from pathlib import Path, PurePath, PurePosixPath
 from random import Random
 
-__all__ = ("ResourcePath",)
+__all__ = ("ResourcePath", "ResourcePathExpression")
```
[the remark this answers is not in the document]
Comment (TallJimbo) [reply python/lsst/resources/_resourcePath.py:29]: I've updated `__init__.py`; thanks for the heads-up.

File: python/lsst/resources/_resourcePath.py [hunk 3]
```diff
@@ -775,7 +823,11 @@ def temporary_uri(
             location exists is the responsibility of the caller.
         suffix : `str`, optional
             A file suffix to be used. The ``.`` should be included in this
-            suffix.
+            suffix.  Passing a string that ends with ``/`` causes a
+            directory-like to be returned even if ``forceDirectory`` is
+            `False`.
+        forceDirectory : `bool`, optional
```
[the remark this answers is not in the document]
Comment (TallJimbo) [reply python/lsst/resources/_resourcePath.py]: It's a failure to back out changes I thought better of - I _did_ try to make this support creating temporary directories, and then realized that cleaning up a temporary remote directory was hairy enough that I didn't want to try without a good reason.

File: python/lsst/resources/_resourcePath.py [hunk 4]
```diff
@@ -1112,3 +1167,86 @@ def findFileResources(
...
+                local_cm = self.as_local()
+            else:
+                local_cm = self.temporary_uri(suffix=self.getExtension())
+            with local_cm as local_uri:
+                with open(local_uri.ospath, mode=mode, encoding=encoding) as file_buffer:
+                    if "a" in mode:
+                        file_buffer.seek(0, io.SEEK_END)
+                    yield file_buffer
+                if "r" not in mode or "+" in mode:
+                    self.transfer_from(local_uri, transfer="copy", overwrite=("x" not in mode))
```
[the remark this answers is not in the document]
Comment (TallJimbo) [reply python/lsst/resources/_resourcePath.py:1237]: Good point; I was indeed assuming that any subclass for which `as_local` could return non-temporary would also override `open`.  I'll add an assert and a note to the docstring to that effect; I think it's a reasonable assumption, but one that should be explicit.

File: tests/test_uri.py (hunk 1 above)
Comment (TallJimbo) [reply tests/test_uri.py]: I think this was another casualty of my aborted attempt to add directory support to `temporary_uri`.  Will restore.

File: python/lsst/resources/packageresource.py [hunk 5]
```diff
@@ -36,3 +38,21 @@ def read(self, size: int = -1) -> bytes:
...
+        prefer_file_temporary: bool = False,
+    ) -> Iterator[IO]:
+        # Docstring inherited.
+        if "r" not in mode or "+" in mode:
+            raise RuntimeError(f"Package resource URI {self} is read-only.")
+        if "b" in mode:
+            with pkg_resources.resource_stream(self.netloc, self.relativeToPathRoot) as buffer:
+                yield buffer
+        else:
+            with super().open(mode, encoding=encoding, prefer_file_temporary=prefer_file_temporary) as buffer:
```
[the remark this answers is not in the document]
Comment (TallJimbo) [reply python/lsst/resources/packageresource.py:57]: Yeah, that would definitely be better, but I'm not about to hold up the ticket for it.
